### PR TITLE
WIP: Add support for keyword values for min-width, min-height, max-width, and max-height

### LINKED ISSUE #13821: Support keyword values for min- and max-width
min-width, max-width, min-block-size, min-inline-height, (and their max- variants) support [a bunch of keyword values](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width#Formal_syntax).

We should handle these in style, layout, and add stylo glue. At the very least we should have feature-parity with Gecko (which implements these keywords prefixed, but implements them nevertheless)

### PR DESCRIPTION
**Pull request by glasserc**
This PR aims to add support for keyword-values `max-content`, `min-content`, `fit-content`, and `fill-available` to the properties that use them, namely `min-width`, `min-height`, `max-width`, and `max-height`.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #13821 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14432)

### CODE REVIEW

File: components/style/gecko/values.rs [hunk 1]
```diff
@@ -187,6 +189,48 @@ impl GeckoStyleCoordConvertible for Angle {
     }
 }
 
+impl GeckoStyleCoordConvertible for ExtremumLength {
+    fn to_gecko_style_coord<T: CoordDataMut>(&self, coord: &mut T) {
+        coord.set_value(CoordDataValue::Enumerated(
+            match *self {
```
Comment (Manishearth) [inline components/style/gecko/values.rs]: Please don't hardcode the constants, instead import them from `gecko_bindings::structs::`

https://github.com/glasserc/servo/blob/413a73f3685600f70303deb5cdee2af0b86b0442/components/style/gecko_bindings/structs_debug.rs#L508

File: components/style/properties/gecko.mako.rs [hunk 2]
```diff
@@ -497,6 +497,7 @@ impl Debug for ${style_struct.gecko_struct_name} {
         "LengthOrPercentage": impl_style_coord,
         "LengthOrPercentageOrAuto": impl_style_coord,
         "LengthOrPercentageOrNone": impl_style_coord,
+        "MinLength": impl_style_coord,
```
Comment (glasserc) [inline components/style/properties/gecko.mako.rs]: @Manishearth told me that I couldn't use `impl_style_coord` any more. Was that meant to be descriptive, or prescriptive? I did have something that compiles so I think I was able to use it, but maybe you meant that `impl_style_coord` isn't meant for this use case or something like that. If I'm not supposed to use `impl_style_coord`, what do I do to cover `min-height` and `min-width`? (Just do it twice? Share a method?)

File: components/style/values/computed/length.rs [hunk 3]
```diff
@@ -473,4 +474,69 @@ pub type LengthOrNone = Either<Length, None_>;
 
 pub type LengthOrNumber = Either<Length, Number>;
 
+#[derive(Copy, Clone, PartialEq)]
```
Comment (glasserc) [inline components/style/values/computed/length.rs]: In order to define how to animate `MinLength`, I tried to copy the style used in `VerticalAlign`, which starts with `match (*self, *other)`. I got compiler errors because this meant moving the `MinLength`. I made the compiler error go away by defining `MinLength` to be `Copy`. Is this right? If not, how do I solve the compiler error for `Interpolate`?

File: components/style/values/computed/length.rs [hunk 4]
```diff
@@ -473,4 +474,69 @@ pub type LengthOrNone = Either<Length, None_>;
...
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum MinLength {
+    LengthOrPercentage(LengthOrPercentage),
+    Auto,
+    ExtremumLength(ExtremumLength),
+}
+
+impl fmt::Debug for MinLength {
```
Comment (Manishearth) [inline components/style/values/computed/length.rs]: I think `derive(Debug)` is fine here

File: components/style/properties/properties.mako.rs [hunk 5]
```diff
@@ -1181,13 +1181,21 @@ impl ComputedValues {
     #[inline]
     pub fn min_inline_size(&self) -> computed::LengthOrPercentage {
         let position_style = self.get_position();
-        if self.writing_mode.is_vertical() { position_style.min_height } else { position_style.min_width }
+        let size = if self.writing_mode.is_vertical() { position_style.min_height } else { position_style.min_width };
+        match size {
+            computed::MinLength::LengthOrPercentage(lop) => lop,
+            _ => panic!("layout doesn't work!")
```
Comment (glasserc) [inline components/style/properties/properties.mako.rs]: Of course, this totally broke because `position_style.min_height` and `position_style.min_width` are both `MinLength` now, instead of `LengthOrPercentage`. I believe that @Manishearth told me in #13821 that I will need to do something to make sure that Servo doesn't parse the keywords. Is that how I'm supposed to handle this case? I'm sure that what I have is completely wrong...

File: components/style/values/mod.rs [hunk 6]
```diff
@@ -161,3 +162,45 @@ impl<A: ToComputedValue, B: ToComputedValue> ToComputedValue for Either<A, B> {
...
+            ExtremumLength::FillAvailable =>
+                dest.write_str("fill-available"),
+        }
+    }
+}
+
+impl Parse for ExtremumLength {
+    fn parse(_context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        match try!(input.next()) {
+            Token::Ident(ref value) if value.eq_ignore_ascii_case("max-content") =>
```
Comment (Manishearth) [inline components/style/values/mod.rs]: Use `match_ignore_ascii_case` here.

File: components/style/values/specified/length.rs [hunk 7]
```diff
@@ -15,6 +15,7 @@ use style_traits::ToCss;
 use style_traits::values::specified::AllowedNumericType;
 use super::{Angle, Number, SimplifiedValueNode, SimplifiedSumNode, Time};
 use values::{CSSFloat, Either, FONT_MEDIUM_PX, HasViewportPercentage, None_};
+use values::{ExtremumLength};
```
Comment (Manishearth) [inline components/style/values/specified/length.rs]: Nit: no braces if it's only one item.

**Review by glasserc — COMMENTED** [review]
This is completely untested because I'm still figuring out how to do that, but it compiled successfully in the `incubator/stylo` repo so I tried to rebase it over and ask my next round of questions.

File: components/style/properties/longhand/position.mako.rs [hunk 8]
```diff
@@ -157,10 +157,8 @@ ${helpers.predefined_type("flex-basis",
 
     // min-width, min-height, min-block-size, min-inline-size
     ${helpers.predefined_type("min-%s" % size,
-                              "LengthOrPercentage",
-                              "computed::LengthOrPercentage::Length(Au(0))",
-                              "parse_non_negative",
-                              needs_context=False,
+                              "MinLength",
```
Comment (Manishearth) [inline components/style/properties/longhand/position.mako.rs]: You should have an `if product == "gecko"` that changes the type of this.

Servo mode should still use LengthOrPercentage. Same for the next line.

This should build in both `./mach build-geckolib` and `./mach build -d` mode.

**Review by Manishearth — COMMENTED** [review]
You have the right idea, some changes need to be made.

File: components/style/values/specified/length.rs (hunk 7 above)
Comment (canaltinova) [inline components/style/values/specified/length.rs]: Yes, also `./mach test-tidy` fails because for this.

File: components/style/properties/properties.mako.rs (hunk 5 above)
Comment (canaltinova) [inline components/style/properties/properties.mako.rs]: You can undo these changes since you don't actually change servo side, like Manishearth said here https://github.com/servo/servo/pull/14432#r90549771

File: components/style/values/specified/length.rs [hunk 9]
```diff
@@ -1019,3 +1020,47 @@ impl LengthOrNumber {
...
+    }
+}
+
+impl Parse for MinLength {
+    fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        input.try(|p| Parse::parse(context, p)).map(MinLength::ExtremumLength)
+            .or_else(|()| input.try(LengthOrPercentage::parse_non_negative).map(MinLength::LengthOrPercentage))
+            .or_else(|()| {
+                match try!(input.next()) {
+                    Token::Ident(ref value) if value.eq_ignore_ascii_case("auto") =>
```
Comment (canaltinova) [inline components/style/values/specified/length.rs]: You can use `match_ignore_ascii_case` in here too.

**Review by canova — COMMENTED** [review]
You need to rebase your commits on top of master. It looks good besides these and Manishearth's comments

**Comment by Wafflespeanut** [comment]
@glasserc Any plans to finish this off?

**Comment by Wafflespeanut** [comment]
Closing this based on inactivity...
